### ISSUE TITLE
FR: Remove producer/consumer; Confusing output

### ISSUE THREAD

**Issue by utdrmac** [tezos-reward-distributor-organization/tezos-reward-distributor#491]
There is no real purpose behind the P/C operation of TRD. It does not make TRD process any faster due to parallelism nor does it improve concurrency. What it does do is present confusion:

```
2021-08-14 13:19:25,125 - producer  - INFO - Calculation report is created at '/home/stuff/.trd/reports/tz1RpNqAzQNaZuxeLSbfEzVEzUWssddsL6Kw/calculations/383.csv'
2021-08-14 13:19:25,125 - producer  - INFO - Reward creation is done for cycle 383, created 1135 rewards.
2021-08-14 13:19:25,126 - producer  - INFO - Run mode ONETIME satisfied. Terminating...
2021-08-14 13:19:25,126 - producer  - DEBUG - Producer returning...
2021-08-14 13:19:25,735 - MainThread - INFO - Interrupted.
2021-08-14 13:19:25,735 - MainThread - INFO - TRD is shutting down...
2021-08-14 13:19:25,735 - MainThread - INFO - --------------------------------------------------------
2021-08-14 13:19:25,736 - MainThread - INFO - Sensitive operations are in progress!
2021-08-14 13:19:25,736 - MainThread - INFO - Please wait while the application is being shut down!
2021-08-14 13:19:25,736 - MainThread - INFO - --------------------------------------------------------
```

TRD remained like this for a full minute. Is TRD stuck? Looks like it is shutting down/stopping. But no, not really. In the background is another thread injecting the multiple batches of payouts.

IMO, we should remove the P/C in favor of a simple payout loop. The P/C is just unnecessary overhead/complexity.

**Comment by utdrmac** [comment]
More on this. When using `-M 3` mode, the messages are confusing as it says TRD is shutting down, but not really. In fact, there are still many batches to process until it can shut down.

```
2022-04-09 19:04:10,231 - consumer0 - INFO - 1276 payments will be done in 5 batches
...
2022-04-09 19:04:14,189 - producer  - INFO - Reward creation is done for cycle 462, created 1571 rewards.
2022-04-09 19:04:14,189 - producer  - INFO - Run mode ONETIME satisfied. Terminating...
2022-04-09 19:04:14,189 - producer  - INFO - Sending KeyboardInterrupt signal.
2022-04-09 19:04:22,746 - MainThread - INFO - Interrupted.
2022-04-09 19:04:22,746 - MainThread - INFO - TRD is shutting down...
2022-04-09 19:04:22,746 - MainThread - INFO - --------------------------------------------------------
2022-04-09 19:04:22,747 - MainThread - INFO - Sensitive operations are in progress!
2022-04-09 19:04:22,747 - MainThread - INFO - Please wait while the application is being shut down!
2022-04-09 19:04:22,747 - MainThread - INFO - --------------------------------------------------------
2022-04-09 19:04:22,747 - MainThread - INFO - Lock file removed!
2022-04-09 19:04:41,480 - consumer0 - INFO - Operation op3DA5U2YgHBSVAsbQo58yZoujdZyz8uHPoYsvZT91MBVdWDUP5 is included
2022-04-09 19:04:41,483 - consumer0 - INFO - Payment of batch 1 succeeded in 1 attempt(s)
2022-04-09 19:04:41,484 - consumer0 - INFO - Payment of batch 2 started
... (TRD continues)
```

**Comment by ericlavoie** [comment]
It's normal there are a few thread/process that must end before it does shut down. This is a message from the main thread/process telling you the shutdown command is received and processing. It will close once the child process are ended, do not kill it. It probably either finishing it's batch payments or waiting for payment confirmation. This is correct, not a bug.

**Comment by utdrmac** [comment]
Yes, I completely understand what is happening. The feature request (FR) is to remove the threaded nature as it is completely unnecessary and outputs confusing information.

**Comment by ericlavoie** [comment]
You want a redesign that is not threaded because the log is too hard to understand? Is that correct?

**Comment by jdsika** [comment]
Maybe improve the log messages and see if there are actual race conditions. Have a second pair of eyes on the threading. Rebuilding the whole thing? .... mäh... IMO it is working as intended with just some confusing output

**Comment by ericlavoie** [comment]
I agree with jdiska, the output is only confusing if you do not understand the threads, maybe a dic section to explain a but, some better output.
Something like "mainthread received shutdown commmand, waiting on child threads to complete before shutting down". Would be more descriptive.

**Comment by utdrmac** [comment]
> You want a redesign that is not threaded because the log is too hard to understand? Is that correct?

I never said hard to understand. I said it was confusing to state the program was exiting/shutting down, when in fact it could be 10m until it stops. I'm also saying that the threaded nature provides absolutely zero benefit to the application's performance as a whole and only adds unnecessary code complexity. I'd estimate a good 500-900 lines of code could be removed and the performance/behavior would be exactly the same.

Changing the messaging would be a good band-aid so that it's no longer confusing.

**Comment by ericlavoie** [comment]
Actually if more than one ops occurs during that span, it does make sense. From what I saw of the code, it would be a fair bit of redesign. There are IMO benefits to the multi threaded design in the way it works. All because the shutdown occurs 10 minutes after it says it is shutting down. Is it overly complex, yeah I agree. But a redesign would be as complex and doing sequential with no thread will bring problems of it's own.

**Comment by nicolasochem** [comment]
:+1: I never understood the benefits of producer/consumer. It's like we are forcing ourselves to do it async processing in what is fundamentally a synchronous path. Step 1: calculate the rewards. Step 2: pay them. Why do we have this in the first place?

I could see paying several cycles in a row, in this case you calculate all cycles in a sequence and you pay all batches in a sequence, in parallel. So you can save time here. But what's the point? Paying several cycles in a row is dangerous and unusual, not something we want to make faster.

**Comment by ericlavoie** [comment]
Maybe it would male sense, if you ran payment for more than one baker on one system (it does not) It could be the reason it was designed that way.

**Comment by habanoz** [comment]
> @habanoz can you please provide historical context on this? Why do we have a producer and a consumer?

It was designed this way from the very beginning. TRD is not designed solely for single baker single cycle payments. Multiple bakers support was on the road map but eventually had to be dropped. (see #15 )

P/C is preferred due to its extensible and scalable nature. Another benefit is concern separation:

**Producers**
As far as I can remember there are three producers: 
1) Reward producer (aka payment producer): There can be more than one pending cycle to pay even for a single baker. Reward calculation can be slow and may benefit from asynchronous behavior. 
2) Retry producer: This one runs parallel to the reward producer and tries to complete failed payments. 
3) Batch producer: This one is provided as part of the batch payment utility. 

**Consumer**
There is only one consumer. Payment consumer only knows how to pay. It is only interested in addresses and amounts. It produces a report of payments attempted.

Regards
